### PR TITLE
Iaas resources

### PR DESCRIPTION
**Pull request by schirevko**

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+---
+layout: "vkcs"
+page_title: "VKCS provider changelog"
+description: |-
+  The VKCS provider's changelog.
+---
+
+# VKCS Provider's changelog
+
+#### v0.5.8
```
Comment (ftersin) [inline docs/changelog.md:10]: lets remove these versions

File: docs/data_sources/vkcs_blockstorage_snapshot.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+---
+layout: "vkcs"
+page_title: "VKCS: blockstorage_snapshot"
+description: |-
+  Get information on an OpenStack Snapshot.
```
Comment (ftersin) [inline docs/data_sources/vkcs_blockstorage_snapshot.md:5]: VKCS Volume Snapshot

File: docs/data_sources/vkcs_blockstorage_volume.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+---
+layout: "vkcs"
+page_title: "VKCS: blockstorage_volume"
+description: |-
+  Get information on an OpenStack Volume.
```
Comment (ftersin) [inline docs/data_sources/vkcs_blockstorage_volume.md:5]: OpenStack -> VKCS, here and everywhere

File: docs/data_sources/vkcs_compute_flavor.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
+---
+layout: "vkcs"
+page_title: "VKCS: compute_flavor_v2"
```
Comment (ftersin) [inline docs/data_sources/vkcs_compute_flavor.md:3]: v2 is not required

File: docs/data_sources/vkcs_keymanager_container.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
+---
+layout: "vkcs"
+page_title: "VKCS: keymanager_container"
+description: |-
+  Get information on a V1 Barbican container resource within OpenStack.
```
Comment (ftersin) [inline docs/data_sources/vkcs_keymanager_container.md:5]: V1 Barbican container resource within OpenStack -> key container resource within VKCS

File: docs/data_sources/vkcs_keymanager_container.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+---
+
+# vkcs\_keymanager\_container
+
+Use this data source to get the ID of an available Barbican container.
```
Comment (ftersin) [inline docs/data_sources/vkcs_keymanager_container.md:10]: Barbican container -> key container

File: docs/data_sources/vkcs_blockstorage_snapshot.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+```hcl
+data "vkcs_blockstorage_snapshot" "snapshot_1" {
+  name        = "snapshot_1"
+  most_recent = true
+}
+```
+
+## Argument Reference
+
+* `region` - (Optional) The region in which to obtain the V3 Block Storage
```
Comment (ftersin) [inline docs/data_sources/vkcs_blockstorage_snapshot.md:23]: d/V3/

File: docs/data_sources/vkcs_blockstorage_volume.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+```hcl
+data "vkcs_blockstorage_volume" "volume_1" {
+  name = "volume_1"
+}
+```
+
+## Argument Reference
+
+* `region` - (Optional) The region in which to obtain the V3 Block Storage
```
Comment (ftersin) [inline docs/data_sources/vkcs_blockstorage_volume.md:22]: d/V3/

File: docs/data_sources/vkcs_compute_flavor.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+```hcl
+data "vkcs_compute_flavor" "small" {
+  vcpus = 1
+  ram   = 512
+}
+```
+
+## Argument Reference
+
+* `region` - (Optional) The region in which to obtain the V2 Compute client.
```
Comment (ftersin) [inline docs/data_sources/vkcs_compute_flavor.md:23]: d/V2/ here and for entire Nova

File: docs/data_sources/vkcs_keymanager_secret.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+  mode        = "cbc"
+  secret_type = "passphrase"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional) The region in which to obtain the V1 KeyManager client.
```
Comment (ftersin) [inline docs/data_sources/vkcs_keymanager_secret.md:31]: d/V1/

File: docs/data_sources/vkcs_keymanager_secret.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
+---
+layout: "vkcs"
+page_title: "VKCS keymanager_secret"
+description: |-
+  Get information on a V1 Barbican secret resource within OpenStack.
```
Comment (ftersin) [inline docs/data_sources/vkcs_keymanager_secret.md:5]: key secret resource within VKCS

File: docs/data_sources/vkcs_keymanager_secret.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+---
+
+# vkcs\_keymanager\_secret
+
+Use this data source to get the ID and the payload of an available Barbican
```
Comment (ftersin) [inline docs/data_sources/vkcs_keymanager_secret.md:10]: Barbican -> VKCS

File: .github.com/workflows/golangci-lint.yml [hunk 13]
```diff
@@ -0,0 +1,84 @@
+
```
Comment (irlndts) [inline .github.com/workflows/golangci-lint.yml:1]: .github.com directory is an artefact, you don't need it, it doesn't affect anything

File: scripts/gogetcookie.sh [hunk 14]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (irlndts) [inline scripts/gogetcookie.sh:1]: the script is unused

+ it looks invalid 
https://github.com/hashicorp/terraform/commit/336b352d6fe4de90392691d892557d5fa2cd07d4

i'm not sure we really need it, maybe we need some research about it.

Comment (schirevko) [reply scripts/gogetcookie.sh]: Removed it

**Comment by dalt74** [comment]
block storage: change availability zone doesn't working

**Comment by dalt74** [comment]
compute: block_device size parameter is called "volume_size". Confused because on block_storage the same is called "size".

**Comment by dalt74** [comment]
compute: block_device image_id accepts only image uuid. The same on block storage accepts image name as well. The semantics supposed to be equal.

**Comment by dalt74** [comment]
compute+blockstorage: define separate volume (VolumeA) from image, then instance with two BDM's - the first for the new volume and the second for the VolumeA. Apply failed with {"badRequest":{"message": "Block Device Mapping is Invalid: Boot sequence for the instance and image/block device mapping combination is not valid.", "code": 400}}

**Comment by dalt74** [comment]
compute: while using delete_on_termination=true for the preexisting volume there is a case for the data loss on instance delete. For the such cases user must confirm he understand  possible data loss with some flags/variables.

**Comment by dalt74** [comment]
build: support for a non-x86 linux arch. Maybe the same for the darwin (M1/aarch64)

**Comment by dalt74** [comment]
keypair: maybe support for pre-existing public keys (?)

**Comment by schirevko** [comment]
> compute: block_device size parameter is called "volume_size". Confused because on block_storage the same is called "size".
Main concept was to take openstack_provider and modify it to support vkcs features so that users who used openstack_provider for vkcs could use our own provider instead. 
This parameters are called the same in openstack provider - "volume_size" for compute and "size" for blockstorage

> compute: while using delete_on_termination=true for the preexisting volume there is a case for the data loss on instance delete. For the such cases user must confirm he understand possible data loss with some flags/variables.

This logic is taken from openstack_provider in accordance to main provider concept. I think, such confirmation through flags/variables can be added in separate task.

> biild: maybe need system-wide install (there is an local-to-user only right now)

Can you elaborate please? End user will not install our provider specifically, this will be done by terraform itself based on user's configuration files. The instruction in README.MD that involves copying provider binaries is related only to development.

> build: support for a non-x86 linux arch. Maybe the same for the darwin (M1/aarch64)

go build will use GOARCH environment variable for this, by default it is amd64, perhaps this is enough?

> keypair: maybe support for pre-existing public keys (?)

Can you elaborate please? User can specify parameter "public_key" and it will be used in resource creation.

File: docs/resources/vkcs_lb_monitor.md [hunk 15]
```diff
@@ -0,0 +1,89 @@
+---
+layout: "vkcs"
+page_title: "vkcs: lb_monitor"
```
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/resources/vkcs_lb_monitor.md:3]: This naming is according to requirements (keep openstack provider names)

File: README.md [hunk 16]
```diff
@@ -0,0 +1,59 @@
+Terraform VKCS Provider
+============================
+
+* Documentation [https://registry.terraform.io/providers/VKCloudSolutions/vkcs/latest/docs](https://registry.terraform.io/providers/VKCloudSolutions/vkcs/latest/docs)
```
[the remark this answers is not in the document]
Comment (schirevko) [reply README.md:4]: Yes, we will change it later when provider is published

File: docs/data_sources/vkcs_networking_router.md [hunk 17]
```diff
@@ -0,0 +1,53 @@
+---
+layout: "vkcs"
+page_title: "vkcs: networking_router"
+description: |-
+  Get information on a VKCS Floating IP.
```
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/data_sources/vkcs_networking_router.md]: Fixed

File: docs/index.md [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+page_title: "Provider: VKCS"
+description: |-
+  The VKCS provider is used to interact with VKCS services.
+  The provider needs to be configured with the proper credentials before it can be used.
+---
+
+# VKCS Provider
+
+The VKCS provider is used to interact with
+[VKCS services](https://vkcs.mail.ru/). The provider needs
```
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/index.md]: Fixed

File: docs/resources/vkcs_lb_loadbalancer.md [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `admin_state_up` - (Optional) The administrative state of the Loadbalancer.
+	A valid value is true (UP) or false (DOWN).
+
+* `availability_zone` - (Optional) The availability zone of the Loadbalancer.
```
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/resources/vkcs_lb_loadbalancer.md:27]: We have data_source_vkcs_compute_availability_zones that can be used to get availability_zones without having to list them manually in doc

File: docs/data_sources/vkcs_networking_network.md [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+* `region` - See Argument Reference above.
+* `external` - See Argument Reference above.
+* `shared` - Specifies whether the network resource can be accessed by any
+   tenant or not.
+* `mtu` - See Argument Reference above.
+* `subnets` - A list of subnet IDs belonging to the network.
+* `all_tags` - The set of string tags applied on the network.
+* `private_dns_domain` - See Argument Reference above.
+* `sdn` - See Argument Reference above.
+* `vkcs_services_access` - Specifies whether VKCS services access is enabled.
```
Comment (Etoneja) [inline docs/data_sources/vkcs_networking_network.md:61]: Has this name (`vkcs_services_access`) been approved by marketing dep? I thought they promoting term 'shadowport' (`enable_shadow_port`).

File: docs/index.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (Etoneja) [reply docs/index.md]: Still exists some in readme. But it cosmetic and (I believe) will be fixed later, so resolve thread.

File: vkcs/resource_vkcs_networking_network.go [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+		log.Printf("[DEBUG] Set tags %s on vkcs_networking_network %s", tags, d.Id())
+	}
+
+	if d.HasChange("vkcs_services_access") {
+		servicesAccess := d.Get("vkcs_services_access").(bool)
+		if !servicesAccess {
+			return diag.Errorf("services_access cannot be disabled")
+		}
+		updateOpts.ServicesAccess = &servicesAccess
+	}
```
Comment (Etoneja) [inline vkcs/resource_vkcs_networking_network.go:252]: Why is this check needed and where did this restriction come from?

File: vkcs/provider.go [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+	if ok {
+		insecure := v.(bool)
+		config.Insecure = &insecure
+	}
+	v, ok = d.GetOk("auth_url")
+	if ok {
+		config.IdentityEndpoint = v.(string)
+	} else {
+		config.IdentityEndpoint = defaultIdentityEndpoint
+	}
```
Comment (Etoneja) [inline vkcs/provider.go:120]: Why no to try take `IdentityEndpoint` from env `OS_AUTH_URL` before setting `defaultIdentityEndpoint` if provider's `auth_url` not specified?

File: vkcs/resource_vkcs_networking_network.go (hunk 21 above)
Comment (Etoneja) [reply vkcs/resource_vkcs_networking_network.go:252]: Right. Restriction exists. Thread closed.

File: docs/data_sources/vkcs_networking_network.md (hunk 20 above)
Comment (schirevko) [reply docs/data_sources/vkcs_networking_network.md:61]: This name came from Mikhail Tikhonov

**Comment by Etoneja** [comment]
Reviewed net and lb resources.
Tested base cases on vkcs custom stuff (sdn flag and shadowport network): works as expected. And the other things seems done with respect for standard openstack api logic.
There is room for improvement, but at the moment, I think it's more important to support all common api features without deep custom validators. And it's done.

So, for me - it's ok.
gj

**Comment by dalt74** [comment]
> > compute: block_device image_id accepts only image uuid. The same on block storage accepts image name as well. The semantics supposed to be equal.
> 
> Terraform does not validate blockstorage's image_id, just passes it to cloud API, that accepts both uuid and image_name.
>

There is a way to resolve image name to ID before call.
 
> > compute+blockstorage: define separate volume (VolumeA) from image, then instance with two BDM's - the first for the new volume and the second for the VolumeA. Apply failed with {"badRequest":{"message": "Block Device Mapping is Invalid: Boot sequence for the instance and image/block device mapping combination is not valid.", "code": 400}}
> 
> Error is caused by boot_index parameter, that should be present if more than one volume is used. Openstack provider would fail with same error if configuration without boot_index is used.

Nova CLI doesn't fail "new volume with bootindex and existing volume without bootindex" but provider does:

$ nova boot --flavor <flavor_id> --nic net-id=<net_id> --availability-zone <zone> --block-device bootindex=0,source=image,dest=volume,id=<image_id>,size=15 --block-device source=volume,dest=volume,id=<volume_id> Server-Name

**Comment by dalt74** [comment]
Assuming build issues may be omitted, the volume service part is working.

The only issue is a compute part.

The volume_size / size naming may be omitted too but the instance build boot_index issue need to be fixed.

**Comment by schirevko** [comment]
@dalt74 
It seems like boot_index issue is out of scope of this particular pull request. Initial task was to take openstack_provider and retool it for vkcs_cloud, removing unnecessary code and adding vkcs-specific code. boot_index is an issue of openstack_provider (main golang library that interacts with openstack cloud uses default value of 0 for this parameter, while Nova CLI does not). This issue may be fixed later, because fixing bugs of openstack_provider was not necessary requirement at this time.

File: vkcs/resource_vkcs_compute_instance.go [hunk 23]
```diff
@@ -0,0 +1,1501 @@
...
+						},
+						"disk_bus": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+			"scheduler_hints": {
```
Comment (gmmephisto) [inline vkcs/resource_vkcs_compute_instance.go:270]: Не уверен что нам нужно экспозить столько хинтов шедулера, особенно то что относится к cell которых у нас и near_host_ip.

File: vkcs/resource_vkcs_compute_instance.go [hunk 24]
```diff
@@ -0,0 +1,1501 @@
...
+				Set:      schema.HashString,
+			},
+			"availability_zone": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				ForceNew:         true,
+				Computed:         true,
+				DiffSuppressFunc: suppressAvailabilityZoneDetailDiffs,
+			},
+			"network_mode": {
```
Comment (gmmephisto) [inline vkcs/resource_vkcs_compute_instance.go:119]: Это какое-то страшное легаси, давайте убирать

File: vkcs/data_source_vkcs_blockstorage_volume.go [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v3/volumes"
+)
+
+func dataSourceBlockStorageVolume() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceBlockStorageVolumeRead,
+
+		Schema: map[string]*schema.Schema{
```
Comment (gmmephisto) [inline vkcs/data_source_vkcs_blockstorage_volume.go:17]: Тут кажется не хватает в атрибутах availability_zone

File: vkcs/resource_vkcs_compute_instance.go (hunk 24 above)
Comment (gmmephisto) [reply vkcs/resource_vkcs_compute_instance.go:119]: А может и не легаси, щас попробую запустить с этой опцией.

File: vkcs/networking_network.go [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+
+type networkExtended struct {
+	networks.Network
+	external.NetworkExternalExt
+	portsecurity.PortSecurityExt
+	dns.NetworkDNSExt
+	policies.QoSPolicyExt
+	provider.NetworkProviderExt
+	PrivateDNSDomainExt
+	ServicesAccess bool `json:"enable_shadow_port,omitempty"`
```
Comment (gmmephisto) [inline vkcs/networking_network.go:31]:
```
diff --git a/vkcs/networking_network.go b/vkcs/networking_network.go
index 26859a0..38bc261 100644
--- a/vkcs/networking_network.go
+++ b/vkcs/networking_network.go
@@ -20,6 +20,10 @@ type PrivateDNSDomainExt struct {
        PrivateDNSDomain string `json:"private_dns_domain,omitempty"`
 }
 
+type ServicesAccessExt struct {
+       ServicesAccess *bool `json:"enable_shadow_port,omitempty"`
+}
+
 type networkExtended struct {
        networks.Network
        external.NetworkExternalExt
@@ -28,7 +32,7 @@ type networkExtended struct {
        policies.QoSPolicyExt
        provider.NetworkProviderExt
        PrivateDNSDomainExt
-       ServicesAccess bool `json:"enable_shadow_port,omitempty"`
+       ServicesAccessExt
 }
 
 // networkingNetworkID retrieves network ID by the provided name.
 ```

Получение ext-net через data атрибут vkcs_networking_networks падает с ошибкой `Error: Unable to retrieve vkcs_networking_networks: json: cannot unmarshal object into Go value of type bool`. Выше diff который фиксит это.

File: vkcs/blockstorage_volume.go [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+		v, err := volumes.Get(client, volumeID).Extract()
+		if err != nil {
+			if _, ok := err.(gophercloud.ErrDefault404); ok {
+				return v, bsVolumeStatusDeleted, nil
+			}
+			return nil, "", err
+		}
+		if v.Status == "error" {
+			return v, v.Status, fmt.Errorf("The volume is in error status. " +
+				"Please check with your cloud admin or check the Block Storage " +
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/blockstorage_volume.go]: Поменял

File: vkcs/compute_instance.go [hunk 28]
```diff
@@ -0,0 +1,476 @@
...
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const (
+	computeAPIMicroVersion = "2.42"
+	// ComputeInstanceCreateServerWithNetworkModeMicroversion = "2.42"
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/compute_instance.go]: Удалил

File: vkcs/resource_vkcs_lb_pool.go [hunk 29]
```diff
@@ -0,0 +1,360 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"protocol": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"TCP", "UDP", "HTTP", "HTTPS", "PROXY", "SCTP", "PROXYV2",
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/resource_vkcs_lb_pool.go]: Убрал

File: vkcs/resource_vkcs_lb_pool.go [hunk 30]
```diff
@@ -0,0 +1,360 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"lb_method": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"ROUND_ROBIN", "LEAST_CONNECTIONS", "SOURCE_IP", "SOURCE_IP_PORT",
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/resource_vkcs_lb_pool.go]: Убрал

File: vkcs/resource_vkcs_lb_listener.go [hunk 31]
```diff
@@ -0,0 +1,468 @@
...
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"protocol": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"TCP", "UDP", "SCTP", "HTTP", "HTTPS", "TERMINATED_HTTPS",
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/resource_vkcs_lb_listener.go]: Убрал

File: vkcs/resource_vkcs_lb_listener.go [hunk 32]
```diff
@@ -0,0 +1,468 @@
...
+	var sniContainerRefs []string
+	if raw, ok := d.GetOk("sni_container_refs"); ok {
+		for _, v := range raw.([]interface{}) {
+			sniContainerRefs = append(sniContainerRefs, v.(string))
+		}
+	}
+
+	var createOpts octavialisteners.CreateOptsBuilder
+	opts := octavialisteners.CreateOpts{
+		// Protocol SCTP requires octavia minor version 2.23
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/resource_vkcs_lb_listener.go]: Убрал

File: vkcs/resource_vkcs_lb_l7policy.go [hunk 33]
```diff
@@ -0,0 +1,420 @@
...
+	if err != nil {
+		return nil, checkDeleted(d, err, "L7 Policy")
+	}
+
+	log.Printf("[DEBUG] Retrieved L7 Policy %s during the import: %#v", d.Id(), l7Policy)
+
+	if l7Policy.ListenerID != "" {
+		d.Set("listener_id", l7Policy.ListenerID)
+	} else {
+		// Fallback for the Neutron LBaaS extension
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/resource_vkcs_lb_l7policy.go]: Убрал

File: vkcs/data_source_vkcs_blockstorage_volume.go (hunk 25 above)
Comment (schirevko) [reply vkcs/data_source_vkcs_blockstorage_volume.go:17]: Добавил

File: vkcs/networking_network.go (hunk 26 above)
Comment (schirevko) [reply vkcs/networking_network.go]: Поправил

File: vkcs/resource_vkcs_compute_instance.go (hunk 23 above)
Comment (gmmephisto) [reply vkcs/resource_vkcs_compute_instance.go:270]: Давай оставим только `group`, а остальные уберем

Comment (schirevko) [reply vkcs/resource_vkcs_compute_instance.go:270]: Done

File: vkcs/resource_vkcs_networking_subnet_test.go [hunk 34]
```diff
@@ -0,0 +1,676 @@
...
+  }
+
+  timeouts {
+    create = "5m"
+    delete = "5m"
+  }
+}
+`
+
+// const testAccNetworkingSubnetPool = `
```
Comment (gmmephisto) [inline vkcs/resource_vkcs_networking_subnet_test.go:459]: Uncomment or remove?

File: vkcs/resource_vkcs_networking_subnet_test.go [hunk 35]
```diff
@@ -0,0 +1,676 @@
...
+  }
+
+  allocation_pool {
+    start = "10.3.0.2"
+    end = "10.3.0.255"
+  }
+}
+`
+
+//const testAccNetworkingSubnetAllocationPool1 = `
```
Comment (gmmephisto) [inline vkcs/resource_vkcs_networking_subnet_test.go:595]: Uncomment or remove?

File: .golangci.yml [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+    - unconvert
+    - unused
+    - varcheck
+    - vet
+    #- gocritic
+    - stylecheck
+
+linters-settings:
+  errcheck:
+    ignore: github.com/hashicorp/terraform-plugin-sdk/helper/schema:ForceNew|Set,fmt:.*,io:Close,github.com/MailRuCloudSolutions/terraform-provider-vkcs:Set
```
Comment (gmmephisto) [inline .golangci.yml:44]: replace `MailRuCloudSolutions` with `vk-cs` ?

File: docs/index.md [hunk 37]
```diff
@@ -0,0 +1,66 @@
...
+
+Terraform 1.0 and later:
+
+```terraform
+# Configure the vkcs provider
+
+terraform {
+  required_providers {
+    vkcs = {
+      source = "MailRuCloudSolutions/vkcs"
```
Comment (gmmephisto) [inline docs/index.md:27]: `MailRuCloudSolutions` -> `vk-cs` ? Also, where we should store releases of provider? Hashiorp registry is not available from RU region.

File: docs/index.md [hunk 38]
```diff
@@ -0,0 +1,66 @@
...
+      version = "~> 0.5.8"
```
Comment (gmmephisto) [inline docs/index.md:28]: Should we start with fresh version such as 0.1.0?

File: docs/index.md [hunk 39]
```diff
@@ -0,0 +1,66 @@
...
+  If omitted, the `USER_NAME` environment variable is used.
+
+* `password` - (Required) The Password to login with. If omitted, the `PASSWORD` environment variable is used.
+
+* `project_id` - (Required) The ID of Project to login with. 
+  If omitted, the `PROJECT_ID` environment variable is used.
+
+* `auth_url` - (Optional) URL for authentication in VKCS. Default is https://infra.mail.ru/identity/v3/.
+
+* `region` - (Optional) A region to use. Default is `RegionOne`. **New since v0.4.0**
```
Comment (gmmephisto) [inline docs/index.md:65]: `New since` is not relevant for VKCS provider

File: docs/guides/getting_started.md [hunk 40]
```diff
@@ -0,0 +1,32 @@
...
+description: |-
+  Getting Started with the VKCS Provider
+---
+
+# Create basic config for VKCS Provider resources
+
+This example shows how to create a simple terraform configuration for creation of VKCS resources.
+
+First, create a Terraform config file named `main.tf`. Inside, you'll want to include the configuration of
+[VKCS Provider](https://registry.terraform.io/providers/MailRuCloudSolutions/vkcs/latest/docs).
```
Comment (gmmephisto) [inline docs/guides/getting_started.md:13]: MailRuCloudSolutions -> vkcs ?

File: README.md [hunk 41]
```diff
@@ -0,0 +1,59 @@
...
+To start improve it grab the repository, build it and install into local registry repository.
+Builds for MacOS, Windows and Linux are available.
+The example is for MacOS.
+```sh
+$ mkdir -p $GOPATH/src/github.com/vk-cs
+$ cd $GOPATH/src/github.com/vk-cs
+$ git clone git@github.com:vk-cs/terraform-provider-vkcs.git
+$ cd $GOPATH/src/github.com/vk-cs/terraform-provider-vkcs
+$ make build_darwin
+$ mkdir -p ~/.terraform.d/plugins/hub.vkcs.mail.ru/repository/vkcs/0.5.8/darwin_amd64/
```
Comment (gmmephisto) [inline README.md:34]: Replace 0.5.8 with actual first version for new provider

File: .goreleaser.yml [hunk 42]
```diff
@@ -0,0 +1,47 @@
...
+    goos:
+      - darwin
+      - freebsd
+      - linux
+      - windows
+    ignore:
+      - goarch: '386'
+        goos: darwin
+    ldflags:
+      - -s -w -X version.ProviderVersion={{.Version}}
```
Comment (gmmephisto) [inline .goreleaser.yml:27]: Where is `ProviderVersion` variable specified?

File: vkcs/data_source_vkcs_networking_subnet_test.go [hunk 43]
```diff
@@ -0,0 +1,294 @@
...
+					testAccCheckNetworkingSubnetDataSourceID("data.vkcs_networking_subnet.subnet_1"),
+					resource.TestCheckResourceAttr(
+						"data.vkcs_networking_subnet.subnet_1", "all_tags.#", "2"),
+				),
+			},
+		},
+	})
+}
+
+// func TestAccNetworkingSubnetDataSource_networkIdAttribute(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/data_source_vkcs_networking_subnet_test.go]: Removed

File: vkcs/data_source_vkcs_networking_subnet_test.go [hunk 44]
```diff
@@ -0,0 +1,294 @@
...
+  cidr = "192.168.199.0/24"
+  network_id = "${vkcs_networking_network.network_1.id}"
+  tags = [
+    "foo",
+    "bar",
+  ]
+}
+`
+
+// const testAccOpenStackNetworkingSubnetDataSourceSubnetWithSubnetPool = `
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/data_source_vkcs_networking_subnet_test.go]: Removed

File: vkcs/data_source_vkcs_networking_subnet_test.go [hunk 45]
```diff
@@ -0,0 +1,294 @@
...
+	return fmt.Sprintf(`
+%s
+
+data "vkcs_networking_subnet" "subnet_1" {
+  gateway_ip = "${vkcs_networking_subnet.subnet_1.gateway_ip}"
+}
+`, testAccOpenStackNetworkingSubnetDataSourceSubnet)
+}
+
+// func testAccOpenStackNetworkingSubnetDataSourceNetworkIDAttribute() string {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/data_source_vkcs_networking_subnet_test.go]: Removed

File: vkcs/import_vkcs_compute_interface_attach_test.go [hunk 46]
```diff
@@ -0,0 +1,24 @@
+package vkcs
+
+// func TestAccComputeV2InterfaceAttachImport_basic(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/import_vkcs_compute_interface_attach_test.go]: Uncommented

File: vkcs/provider_test.go [hunk 47]
```diff
@@ -0,0 +1,352 @@
...
+		"OS_EXTGW_ID":    osExtGwID,
+	}
+	for k, v := range vars {
+		if v == "" {
+			t.Fatalf("'%s' must be set for acceptance test", k)
+		}
+	}
+}
+
+// func TestProvider(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/provider_test.go]: Removed, will return this with old provider resources

File: vkcs/resource_vkcs_compute_instance_test.go [hunk 48]
```diff
@@ -0,0 +1,1611 @@
...
+					resource.TestCheckResourceAttr(
+						"vkcs_compute_instance.instance_1", "power_state", "active"),
+					testAccCheckComputeInstanceState(&instance, "active"),
+				),
+			},
+		},
+	})
+}
+
+// func TestAccComputeInstance_secgroupMulti(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/resource_vkcs_compute_instance_test.go]: Removed

File: vkcs/resource_vkcs_networking_subnet_test.go (hunk 34 above)
Comment (schirevko) [reply vkcs/resource_vkcs_networking_subnet_test.go]: Removed

File: vkcs/resource_vkcs_networking_subnet_test.go (hunk 35 above)
Comment (schirevko) [reply vkcs/resource_vkcs_networking_subnet_test.go]: Removed

File: .golangci.yml (hunk 36 above)
Comment (schirevko) [reply .golangci.yml]: Replaced

File: docs/index.md (hunk 37 above)
Comment (schirevko) [reply docs/index.md]: Replaced

File: docs/index.md (hunk 38 above)
Comment (schirevko) [reply docs/index.md]: Updated
Some old examples in docs/examples may use old version, they will be updated with old provider resources pull request

File: docs/index.md (hunk 39 above)
Comment (schirevko) [reply docs/index.md]: Removed

File: docs/guides/getting_started.md (hunk 40 above)
Comment (schirevko) [reply docs/guides/getting_started.md]: Replaced

File: README.md (hunk 41 above)
Comment (schirevko) [reply README.md]: Replaced

File: .goreleaser.yml (hunk 42 above)
Comment (schirevko) [reply .goreleaser.yml:27]: It is taken from tag which we provide for provider release

**Review by gmmephisto — APPROVED** [review]

**Pull request merged**